File: demos/01-container-runtime/apps/go/server_test.go
package main

import (
	"encoding/json"
	"io"
	"log/slog"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func testServer(t *testing.T) *server {
	t.Helper()
	cfg := config{
		Port:           8080,
		ServiceName:    "demo-go-api",
		ServiceVersion: "0.1.0",
		LogLevel:       "info",
		Env:            "development",
	}
	return newServer(cfg, slog.New(slog.NewTextHandler(io.Discard, nil)))
}

func TestHealthEndpoints(t *testing.T) {
	srv := testServer(t)
	handler := srv.routes()

	tests := []struct {
		name string
		path string
	}{
		{name: "live", path: "/health/live"},
		{name: "ready", path: "/health/ready"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			req := httptest.NewRequest(http.MethodGet, tt.path, nil)
			rec := httptest.NewRecorder()
			handler.ServeHTTP(rec, req)

			if rec.Code != http.StatusOK {
				t.Fatalf("status = %d, want 200", rec.Code)
			}
			if ct := rec.Header().Get("Content-Type"); ct != "application/json" {
				t.Fatalf("Content-Type = %q, want application/json", ct)
			}

			var body healthResponse
			if err := json.NewDecoder(rec.Body).Decode(&body); err != nil {
				t.Fatalf("decode: %v", err)
			}
			if body.Status != "ok" {
				t.Fatalf("status field = %q, want ok", body.Status)
			}
		})
	}
}

func TestIdentityEndpoint(t *testing.T) {
	srv := testServer(t)
	// Freeze uptime > 0
	srv.startedAt = time.Now().UTC().Add(-2 * time.Second)
	handler := srv.routes()

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)

	if rec.Code != http.StatusOK {
		t.Fatalf("status = %d, want 200", rec.Code)
	}

	var body identityResponse
	if err := json.NewDecoder(rec.Body).Decode(&body); err != nil {
		t.Fatalf("decode: %v", err)
	}

	if body.Service != "demo-go-api" {
		t.Errorf("service = %q, want demo-go-api", body.Service)
	}
	if body.Language != "go" {
		t.Errorf("language = %q, want go", body.Language)
	}
	if body.Status != "running" {
		t.Errorf("status = %q, want running", body.Status)
	}
	if body.Version != "0.1.0" {
		t.Errorf("version = %q, want 0.1.0", body.Version)
	}
	if body.UptimeSeconds <= 0 {
		t.Errorf("uptime_seconds = %v, want > 0", body.UptimeSeconds)
	}
}

func TestLoadConfigRequiresPort(t *testing.T) {
	t.Setenv("PORT", "")
	t.Setenv("FORGE_LOG_LEVEL", "info")
	if _, err := loadConfig(); err == nil {
		t.Fatal("expected error for missing PORT")
	}
}

func TestLoadConfigRejectsInvalidPort(t *testing.T) {
	t.Setenv("PORT", "not-a-port")
	t.Setenv("FORGE_LOG_LEVEL", "info")
	if _, err := loadConfig(); err == nil {
		t.Fatal("expected error for invalid PORT")
	}
}

func TestLoadConfigRejectsInvalidLogLevel(t *testing.T) {
	t.Setenv("PORT", "8080")
	t.Setenv("FORGE_LOG_LEVEL", "verbose")
	if _, err := loadConfig(); err == nil {
		t.Fatal("expected error for invalid FORGE_LOG_LEVEL")
	}
}

File: demos/01-container-runtime/apps/kotlin/src/test/kotlin/demo/ServerTest.kt
package demo

import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.http.HttpStatusCode
import io.ktor.serialization.kotlinx.json.json
import io.ktor.server.testing.testApplication
import kotlinx.serialization.json.Json
import java.util.concurrent.atomic.AtomicLong
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class ServerTest {
    private val cfg = Config(
        port = 8080,
        serviceName = "demo-kotlin-api",
        serviceVersion = "0.1.0",
        logLevel = "info",
        env = "test",
    )

    @Test
    fun liveAndReady() = testApplication {
        application {
            configureContractRoutes(cfg)
        }
        val client = createClient {
            install(ContentNegotiation) {
                json(Json { ignoreUnknownKeys = true })
            }
        }
        for (path in listOf("/health/live", "/health/ready")) {
            val response = client.get(path)
            assertEquals(HttpStatusCode.OK, response.status)
            val body = response.body<HealthResponse>()
            assertEquals("ok", body.status)
        }
    }

    @Test
    fun identity() = testApplication {
        val startedAt = AtomicLong(System.currentTimeMillis() - 2_000)
        application {
            configureContractRoutes(cfg, startedAt)
        }
        val client = createClient {
            install(ContentNegotiation) {
                json(Json { ignoreUnknownKeys = true })
            }
        }
        val response = client.get("/")
        assertEquals(HttpStatusCode.OK, response.status)
        val body = response.body<IdentityResponse>()
        assertEquals("demo-kotlin-api", body.service)
        assertEquals("kotlin", body.language)
        assertEquals("running", body.status)
        assertEquals("0.1.0", body.version)
        assertTrue(body.uptimeSeconds != null && body.uptimeSeconds!! > 0.0)
    }

    @Test
    fun loadConfigRequiresPort() {
        assertFailsWith<IllegalArgumentException> {
            loadConfig(mapOf("FORGE_LOG_LEVEL" to "info"))
        }
    }

    @Test
    fun loadConfigRejectsInvalidPort() {
        assertFailsWith<IllegalArgumentException> {
            loadConfig(mapOf("PORT" to "not-a-port", "FORGE_LOG_LEVEL" to "info"))
        }
    }

    @Test
    fun loadConfigRejectsInvalidLogLevel() {
        assertFailsWith<IllegalArgumentException> {
            loadConfig(mapOf("PORT" to "8080", "FORGE_LOG_LEVEL" to "verbose"))
        }
    }

    @Test
    fun loadConfigDefaults() {
        val loaded = loadConfig(mapOf("PORT" to "8080"))
        assertEquals(8080, loaded.port)
        assertEquals("demo-kotlin-api", loaded.serviceName)
        assertEquals("info", loaded.logLevel)
    }
}
